Google.Api.Gax.Grpc/ClientBuilderBase.cs

```diff
@@ -190,6 +198,25 @@ protected async virtual Task<ChannelCredentials> GetChannelCredentialsAsync(Canc
         /// </summary>
         protected abstract ServiceEndpoint GetDefaultEndpoint();
 
+        /// <summary>
+        /// Returns the channel pool to use when no other options are specified. This method is not called unless
+        /// <see cref="CanUseChannelPool"/> returns true, so if a channel pool is unavailable, override that property
+        /// to return false and throw an exception from this method.
+        /// </summary>
+        protected abstract ChannelPool GetChannelPool();
+
+        /// <summary>
+        /// Returns whether or not a channel pool can be used if a channel is required. The default behavior is to return
+        /// true if and only if no scopes, credentials or token access method have been specified. Derived classes should
+        /// override this property if there are other reasons why the channel pool should not be used.
+        /// </summary>
+        protected virtual bool CanUseChannelPool =>
```

Comment:
More than `CanUseChannelPool` I would name it `ShouldUseChannelPool` or `NeedsToUseChannelPool`. I mean, if there are no credentials whatsoever, then using the channel pool is the only way.

Reply:
No, it would use the default credentials (which is what the channel pool does anyway).

Happy to consider other names, but I do think of it more as "permission" than "urging" - because it doesn't bother checking for things like whether there's already a CallInvoker etc.

Reply:
Oh yes, that's true, we do have the default credentials for when nothing is set, and we are not enforcing that some credentials need to be set anyways. That mixed with the fact that custom implementations might decide to ignore the ChannelPool entirely, I now agree with the "permission" vs. the "urgin".